File: web/converters/excel.js
// Excel (.xlsx/.xlsm) converter using SheetJS (global XLSX).
// Yields blocks split by sheet and by row-window so large sheets chunk cleanly.

function escapePipe(v) {
  if (v === null || v === undefined) return "";
  return String(v).replace(/\|/g, "\\|").replace(/\r?\n/g, " ");
}

function rowsToMarkdownTable(header, rows) {
  if (!header || header.length === 0) return "";
  const head = "| " + header.map(escapePipe).join(" | ") + " |";
  const sep  = "|" + header.map(() => " --- ").join("|") + "|";
  const body = rows
    .map(r => "| " + header.map((_, i) => escapePipe(r[i])).join(" | ") + " |")
    .join("\n");
  return [head, sep, body].filter(Boolean).join("\n");
}

function sheetToBlocks(wb, sheetName, rowsPerSection) {
  const ws = wb.Sheets[sheetName];
  if (!ws) return [];
  // header: true → rows come back as arrays; includes empty cells as ""
  const aoa = XLSX.utils.sheet_to_json(ws, { header: 1, defval: "", blankrows: false });
  if (aoa.length === 0) {
    return [{
      kind: "sheet-section",
      md: `## 시트: ${sheetName}\n\n_(빈 시트)_\n`,
      meta: { source: `시트: ${sheetName}` },
    }];
  }
  const header = aoa[0];
  const rest = aoa.slice(1);
  const blocks = [];
  // One intro block per sheet plus row-window blocks.
  blocks.push({
    kind: "heading",
    md: `## 시트: ${sheetName}`,
    meta: { source: `시트: ${sheetName}` },
  });
  if (rest.length === 0) {
    blocks.push({
      kind: "sheet-section",
      md: rowsToMarkdownTable(header, []) + "\n",
      meta: { source: `시트: ${sheetName} · 헤더만` },
    });
    return blocks;
  }
  const step = Math.max(20, rowsPerSection | 0 || 200);
  for (let i = 0; i < rest.length; i += step) {
    const slice = rest.slice(i, i + step);
    const from = i + 2;                // +1 for 1-based, +1 for header row
    const to = i + 1 + slice.length;
    const label = `시트: ${sheetName} · 행 ${from}-${to}`;
    const table = rowsToMarkdownTable(header, slice);
    blocks.push({
      kind: "sheet-section",
      md: `### ${label}\n\n${table}\n`,
      meta: { source: label },
    });
  }
  return blocks;
}

export async function convertExcel(file, opts) {
  const buf = await file.arrayBuffer();
  // cellFormula keeps formulas accessible if later needed; cellDates normalizes dates.
  const wb = XLSX.read(buf, { type: "array", cellDates: true, cellNF: false });
  const rowsPerSection = opts?.sheetRows || 200;
  const blocks = [];
  blocks.push({
    kind: "heading",
    md: `# ${file.name}\n\n> Excel 문서 · 시트 ${wb.SheetNames.length}개\n`,
    meta: { source: file.name },
  });
  for (const name of wb.SheetNames) {
    blocks.push(...sheetToBlocks(wb, name, rowsPerSection));
  }
  return { blocks, images: [] };
}

if (typeof window !== "undefined") {
  window.__converters = window.__converters || {};
  window.__converters.excel = convertExcel;
}

File: code/converter/readers/excel.py
"""Excel reader (.xlsx/.xlsm) using openpyxl in read_only/stream mode."""

from __future__ import annotations

from pathlib import Path
from typing import Iterable

from openpyxl import load_workbook

from .common import Block, ImageAsset


def _escape(v) -> str:
    if v is None:
        return ""
    return str(v).replace("|", "\\|").replace("\r\n", " ").replace("\n", " ")


def _row_to_md(headers: list[str], row: list) -> str:
    return "| " + " | ".join(_escape(c) for c in (row + [""] * (len(headers) - len(row)))[: len(headers)]) + " |"


def _header_md(headers: list[str]) -> str:
    head = "| " + " | ".join(_escape(h) for h in headers) + " |"
    sep = "|" + "|".join(" --- " for _ in headers) + "|"
    return head + "\n" + sep


def read(path: str | Path, rows_per_section: int = 200) -> tuple[list[Block], list[ImageAsset]]:
    path = Path(path)
    wb = load_workbook(filename=path, read_only=True, data_only=True)
    blocks: list[Block] = [
        Block(
            kind="heading",
            md=f"# {path.name}\n\n> Excel 문서 · 시트 {len(wb.sheetnames)}개\n",
            source=path.name,
        )
    ]
    for sheet_name in wb.sheetnames:
        ws = wb[sheet_name]
        rows_iter: Iterable[tuple] = ws.iter_rows(values_only=True)
        try:
            headers = [_escape(c) if c is not None else "" for c in next(rows_iter)]
        except StopIteration:
            blocks.append(
                Block(
                    kind="sheet-section",
                    md=f"## 시트: {sheet_name}\n\n_(빈 시트)_\n",
                    source=f"시트: {sheet_name}",
                )
            )
            continue

        if not headers:
            headers = [""]

        blocks.append(Block(kind="heading", md=f"## 시트: {sheet_name}", source=f"시트: {sheet_name}"))

        buf: list[list] = []
        first_row_no = 2
        row_no = first_row_no

        def _flush() -> None:
            nonlocal buf, first_row_no
            if not buf:
                return
            label = f"시트: {sheet_name} · 행 {first_row_no}-{first_row_no + len(buf) - 1}"
            body = "\n".join(_row_to_md(headers, list(r)) for r in buf)
            md = f"### {label}\n\n{_header_md(headers)}\n{body}\n"
            blocks.append(Block(kind="sheet-section", md=md, source=label))
            first_row_no = first_row_no + len(buf)
            buf = []

        for row in rows_iter:
            buf.append(list(row))
            row_no += 1
            if len(buf) >= rows_per_section:
                _flush()
        _flush()
        if first_row_no == 2 and not buf:
            # only header was present
            blocks.append(
                Block(
                    kind="sheet-section",
                    md=f"{_header_md(headers)}\n_(데이터 행 없음)_\n",
                    source=f"시트: {sheet_name} · 헤더만",
                )
            )
    wb.close()
    return blocks, []
